File: 20-Valid-Parentheses.js
var isValid = function(s) {
    const bracketMap = {
        '(': ')',
        '{': '}',
        '[': ']'
    };
    const stack = [s[0]];
    
    for (let i = 1; i < s.length; i++) {
        if (bracketMap[stack[stack.length - 1]] === s[i]) {
            stack.pop();
        } else {
            stack.push(s[i]);
        }
    }
    
    if (!stack.length) {
        return true;
    } else {
        return false;
    }
};

File: go/20-Valid-Parentheses.go
func isValid(s string) bool {
	stack := make([]rune, 0)
	matchingBrackets := map[rune]rune{
		')': '(',
		']': '[',
		'}': '{',
	}
	for _, bracket := range s {
		switch bracket {
		case '(', '{', '[':
			stack = append(stack, bracket)
		case ')', '}', ']':
			if len(stack) == 0 || stack[len(stack)-1] != matchingBrackets[bracket] {
				return false
			}

			stack = stack[:len(stack)-1]
		}
	}

	return len(stack) == 0
}